### PR TITLE
Avoid double-resyncs without leader election.

### PR DESCRIPTION
**Pull request by mattmoor**
tl;dr Without leader election enabled, we've been suffering from double resyncs, this fixes that.

Essentially when the informers are started, prior to starting the controllers, they fill the controller's workqueue with ~the world.  Once we start the controllers, the leader election code calls `Promote` on a `Bucket` so that it starts processing them as their leader, and then is requeues every key that falls into that `Bucket`.

When leader election is disabled, we use the `UniversalBucket`, which always owns everything.  This means that the first pass through the workqueue was already hitting every key, so re-enqueuing every key results in superfluous processing of those keys.

This change makes the `LeaderAwareFuncs` elide the call to `PromoteFunc` when the `enq` function passed is `nil`, and makes the `unopposedElector` have a `nil` `enq` field when we explicitly pass it the `UniversalBucket` because leader election is disabled.

/kind cleanup

**Release Note**

```release-note

```

**Docs**

```docs

```

### CODE REVIEW

**Comment by mattmoor** [comment]
`TestMetricsExport/OpenCensus` in the actions leg that failed

**Comment by dprotaso** [comment]
Can we add some unit tests for the `leaderelection` and `reconciler` package changes?

**Comment by mattmoor** [comment]
Done, RFAL.

File: controller/controller_test.go [hunk 1]
```diff
@@ -968,13 +968,10 @@ func TestStartAndShutdownWithLeaderAwareNoElection(t *testing.T) {
...
 
 	select {
-	case <-promoted:
-		// We expect to be promoted immediately, since there is no
-		// ElectorBuilder attached to the context.
 	case <-doneCh:
 		t.Fatal("StartAll finished early.")
 	case <-time.After(10 * time.Second):
-		t.Error("Timed out waiting for StartAll.")
+		// Give it some time to run.
```
Comment (n3wscott) [inline controller/controller_test.go:974]: the loss of the error message might be a problem? This will fall through to the cancel at line 977 and you might not understand if or why it timed out.

**Review by n3wscott — COMMENTED** [review]
one nit/question, otherwise LGTM

File: controller/controller_test.go (hunk 1 above)
Comment (mattmoor) [reply controller/controller_test.go:974]: This is now the normal path for the test.  Previously `PromoteFunc` was called, which triggered another path.  Now this is waiting to effectively make sure it isn't called.
